compiler_opt/rl/log_reader_test.py

```diff
@@ -30,83 +31,114 @@ def json_to_bytes(d) -> bytes:
   return json.dumps(d).encode('utf-8')
 
 
-nl = '\n'.encode('utf-8')
-
-
-def write_buff(f: BinaryIO, buffer: list, ct):
-  # we should get the ctypes array to bytes for pytype to be happy.
-  f.write((ct * len(buffer))(*buffer))  # pytype:disable=wrong-arg-types
-
-
-def write_context_marker(f: BinaryIO, name: str):
-  f.write(json_to_bytes({'context': name}))
-  f.write(nl)
-
-
-def write_observation_marker(f: BinaryIO, obs_idx: int):
-  f.write(json_to_bytes({'observation': obs_idx}))
-  f.write(nl)
-
-
-def write_nl(f: BinaryIO):
-  f.write(nl)
-
-
-def write_outcome_marker(f: BinaryIO, obs_idx: int):
-  f.write(json_to_bytes({'outcome': obs_idx}))
-  f.write(nl)
-
-
-def create_example(fname: str, nr_contexts=1):
+class LogTestExampleBuilder:
+  """Construct a log."""
+
+  newline = b'\n'
+  error_newline = b'hi there'
+
+  class ErrorMarkers(enum.IntEnum):
```

Comment:
newline, error_newline, and ErrorMarkers are all defined inside LogTestExampleBuilder so any instance of LogTestExampleBuilder can access them via `self.` which is shorter to write/read but will automatically update if the outer class is renamed and they will also be accessible from subclasses of the outer class.

Reply:
Hmm... (being of a C++ mindset) not sure how to digest the auto-update part of the motivation. Is that (friendliness to opportunistic extensibility) the pythonic way though? (happy to follow that! basically asking to educate myself)